receiver/googlecloudspannerreceiver/internal/statsreader/intervalstatsreader.go

```diff
@@ -0,0 +1,149 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package statsreader
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"go.opentelemetry.io/collector/model/pdata"
+	"go.uber.org/zap"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/receiver/googlecloudspannerreceiver/internal/datasource"
+	"github.com/open-telemetry/opentelemetry-collector-contrib/receiver/googlecloudspannerreceiver/internal/metadata"
+)
+
+const (
+	backFillIntervalDuration = time.Hour
+)
+
+type intervalStatsReader struct {
+	currentStatsReader
+	backFillEnabled   bool
+	lastPullTimestamp time.Time
+}
+
+func newIntervalStatsReaderWithMaxRowsLimit(
+	logger *zap.Logger,
+	database *datasource.Database,
+	metricsMetadata *metadata.MetricsMetadata,
+	backFillEnabled bool,
+	topMetricsQueryMaxRows int) *intervalStatsReader {
+
+	reader := currentStatsReader{
+		logger:                 logger,
+		database:               database,
+		metricsMetadata:        metricsMetadata,
+		topMetricsQueryMaxRows: topMetricsQueryMaxRows,
+		statement:              intervalStatsStatement,
+	}
+
+	return &intervalStatsReader{
+		currentStatsReader: reader,
+		backFillEnabled:    backFillEnabled,
+	}
+}
+
+func newIntervalStatsReader(
+	logger *zap.Logger,
+	database *datasource.Database,
+	metricsMetadata *metadata.MetricsMetadata,
+	backFillEnabled bool) *intervalStatsReader {
+
+	reader := currentStatsReader{
+		logger:          logger,
+		database:        database,
+		metricsMetadata: metricsMetadata,
+		statement:       intervalStatsStatement,
+	}
+
+	return &intervalStatsReader{
+		currentStatsReader: reader,
+		backFillEnabled:    backFillEnabled,
+	}
+}
+
+func (reader *intervalStatsReader) Read(ctx context.Context) ([]pdata.Metrics, error) {
+	reader.logger.Info(fmt.Sprintf("Executing read method for reader %v", reader.Name()))
+
+	// Generating pull timestamps
+	pullTimestamps := pullTimestamps(reader.lastPullTimestamp, reader.backFillEnabled)
+
+	var collectedMetrics []pdata.Metrics
+
+	// Pulling metrics for each generated pull timestamp
+	for _, pullTimestamp := range pullTimestamps {
+		stmtArgs := statementArgs{
+			query:                  reader.metricsMetadata.Query,
+			topMetricsQueryMaxRows: reader.topMetricsQueryMaxRows,
+			pullTimestamp:          pullTimestamp,
+		}
+
+		stmt := reader.statement(stmtArgs)
+		metrics, err := reader.pull(ctx, stmt)
+
+		if err != nil {
+			return nil, err
+		}
+
+		collectedMetrics = append(collectedMetrics, metrics...)
+	}
+
+	reader.lastPullTimestamp = pullTimestamps[len(pullTimestamps)-1]
+
+	return collectedMetrics, nil
+}
+
+// This slice will always contain at least one value.
+func pullTimestamps(lastPullTimestamp time.Time, backFillEnabled bool) []time.Time {
```

Comment:
With 9 tables and 60 minute data per table, we will generate 540 queries in a short time in a burst. Is there a way to smoothen this out. 

Other option is to make it as  stale read query. It will get load balanced to other servers and not just the owner of the data.

Reply:
We need to think about this. For now I did it in such way because of applying max rows limitation per timestamp.
What doe is mean stale read query? Can you provide more details about this?

Reply:
@balachandr , Ok, we have 3 options here:
- stale read query
- generate union query of 60 selects(we also looked at partition by option, but looks like Spanner doesn't support it)
- leave it as is
will work on this in scope of separate MR/task.

Reply:
@balachandr, looks like union is not an option - so, only 2 remaining. Will try to resolve this in one of upcoming MRs.